Add form to create a new ticket

mitogh opened the pull request:
Add UI for the form required to create a new ticket inside of the tickets blocks, separate components into smaller
parts to decrease complexity and have smaller tests.

Components part of the ticket form: 

- `<Header />`
- `<Capacity />`
- `<AdvancedOptions />`
- `<AttendeesRegistration />`

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/attendees-registration/template.js:
```diff
@@ -0,0 +1,48 @@
+		attendeeButtonClick,
+	} = props;
+
+	return (
```
this is OK for now, but in https://github.com/moderntribe/events-gutenberg/pull/322 I added a component called `labelWithModal`, which you can call like so:

https://github.com/moderntribe/events-gutenberg/pull/322/files#diff-954eb9025f65e7bcf8ec7eb67c9927b6R24

mitogh replied:
Well I think we are creating components to specifc at this point at least based on names instead we should compose them as: 

- Label -> tooltip( label ) -> modal( tooltip( label ) )

As what we are doing at this point is: 

- Label
- LabelWithToolTip
- LabelWithModal 

First approach seems more scalable, however I haven't look into the labelWithTooltip but seems like we should do a composition instead of specificity on those cases.

paulmskim replied:
I think you might be misunderstanding what the modal portion is, it's actually not a wrapper around the tooltip, it's a replacement of the tooltip. So if you look into that PR, we have a wrapper element called `LabeledItem` which wraps the `children` passed into the `LabeledItem` component with a label next to it. This seems to be a recurring pattern with the latest being the `Tooltip` and `ModalButton`. The labeled elements `LabelWithTooltip` and `LabelWithModal` created from the `Tooltip` and `ModalButton` elements are use cases of `LabeledItem`, which still provide a lot of flexibility (label text, modal button text, modal button functionality, tooltip text, tooltip label, etc.), so I disagree that this is too specific. In fact, these elements (`LabelWithTooltip`, `LabelWithModal`) are composed elements themselves, they just have a name to them.

mitogh replied:
Haven't look into the specifics on this one but is my interpretation as we have similar names as mentioned above.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/capacity/template.js:
```diff
@@ -0,0 +1,153 @@
+};
+
+// Custom input for this type of form
+const Input = ( { type, input, selected } ) => (
```
doesn't have to be in this PR, but would be nice to make this a general component as there are many pieces that can be reused.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/capacity/template.js:
```diff
@@ -0,0 +1,153 @@
+		<div className="tribe-editor__tickets-form__row">
+			<div className="tribe-editor__tickets-form__labels">
+				<LabelWithTooltip
+					id="tickets-capacity-type"
```
This will not work well once you have 2 `ticket` blocks, which we will definitely see. We cannot have 2 or more elements with the same id. Having the same id for all ticket capacity selects will cause problems on the page.

Instead, I suggest 2 things:
- Make this a stateful component by extending `PureComponent`, create a property `this.selectId = uniqid()` in the `constructor`, and add it to the `Select` below.
- As mentioned above with the `LabelWithTooltip` element, as this label and tooltip reference a group of form elements, it's probably better to have a label associated with the select itself, but visually hidden. So something like:

```js
...
  <label htmlFor={ this.selectId } >...</label>
  <Select
    id={ this.selectId }
    ...
  />
...
```

mitogh replied:
Great, not sure where the scenario of 2 ticket blocks comes into place tho would you mind explain that one a bit more?

paulmskim replied:
within the `tickets` block you can have many `ticket` blocks.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/duration/style.pcss:
```diff
@@ -0,0 +1,15 @@
+		margin-left: 0;
+	}
+
+	.tribe-editor__date-time-range-picker__separator--time-range {
```
wonder why this style is needed? doesn't the date time range picker style not have the styles set already?

mitogh replied:
Styles are applied to the container instead as we have an element that is absolute placed inside of this container as a reference.

paulmskim replied:
This is not resolved, this relates to https://github.com/moderntribe/events-gutenberg/pull/319#discussion_r218719427

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/duration/style.pcss:
```diff
@@ -0,0 +1,15 @@
+	position: relative;
+	margin-bottom: 20px;
+
+	.tribe-editor__date-time-range-picker__start {
```
this style should already be taken care of in the date time range picker element. if this is to align the date time range picker with the label + tooltip, check out these:

- https://github.com/moderntribe/events-gutenberg/blob/feature/event-ticket-blocks/plugins/tickets/src/modules/blocks/rsvp/rsvp-duration/template.js#L14
- https://github.com/moderntribe/events-gutenberg/blob/feature/event-ticket-blocks/plugins/tickets/src/modules/blocks/rsvp/rsvp-duration/style.pcss#L1
- https://github.com/moderntribe/events-gutenberg/blob/feature/event-ticket-blocks/plugins/tickets/src/modules/blocks/rsvp/rsvp-duration-label/style.pcss#L1
- https://github.com/moderntribe/events-gutenberg/blob/feature/event-ticket-blocks/plugins/tickets/src/modules/blocks/rsvp/rsvp-duration-picker/template.js
- https://github.com/moderntribe/events-gutenberg/blob/feature/event-ticket-blocks/plugins/tickets/src/modules/blocks/rsvp/rsvp-duration-picker/style.pcss#L1

If you have question let me know and we can check it out together, the elements should really just be plug-and-play.

mitogh replied:
I think the main difference here is the structure of the layout of the inputs as those in the tickets have the same width but the `to` is placed outside of that container.

So we need to make sure the inputs on those fill all the available space on that row, and place the `to` outside of that container.

- http://p.tri.be/7mfKOO

Let me know if you have more questions in particular about that one.

paulmskim replied:
If you just plug in the `<DateTimeRangePicker />` component with no external styling, you will see that that styling is already take care of. Check out below if you want more details on the markup and styles:

https://github.com/moderntribe/events-gutenberg/tree/feature/event-ticket-blocks/plugins/tickets/src/modules/elements/date-time-range-picker

These elements really are plug-and-play, not a ton of styles changes needed, only layout.

paulmskim replied:
The RSVP block has exactly the same feature, which is why I'm saying these styles are unnecessary. As mentioned before, these elements should just be used as is, and fix the outer element layouts as necessary without touching the inner element layout (which should be handled in the element styles). However, ass it is, it's not functionally a blocker, so I will leave it to you to decide if you would like to fix it or not.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/duration/template.js:
```diff
@@ -0,0 +1,56 @@
+	const { label, tooltip } = props;
+	return (
+		<div className="tribe-editor__tickets-form__row">
+			<div className="tribe-editor__tickets-form__labels">
```
as in the previous comment, we shouldn't need wrapper divs for the label and date time range picker, these elements should just be plug and play, check out:

- https://github.com/moderntribe/events-gutenberg/blob/feature/event-ticket-blocks/plugins/tickets/src/modules/blocks/rsvp/rsvp-duration/template.js#L14

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/advanced-options/template.js:
```diff
@@ -0,0 +1,82 @@
+		accordionTitle: __( 'Advanced Options', 'events-gutenberg' ),
+		onClick: () => {},
+	};
+
```
**EDIT**: I didn't realize this was actually an element. This really isn't an element, it's specific for the `ticket` block. I cannot reuse this for RSVP, and there won't be any other blocks/components that would use this element, it's too specific. All of this code under the `ticket-form` should actually be under the `ticket` block as its own components.

mitogh replied:
I realize this after I check on your side, but wanted to place all elements to avoid having nested or specifc ones in directories of each block.

I'm fine with moving them into the `tickets` directory, I think Neal was using `components/` instead of `elements` I think we should define how this should be approached, but I do agree with the part of `elements` should contain only reusable items.

Will create a different PR to avoid more complexity on this one after merging this one by moving all this inside of the `tickets` block.

paulmskim replied:
OK, the sooner the better, as we should architecturally have things in place first.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/footer/style.pcss:
```diff
@@ -0,0 +1,28 @@
+	background-color: #009FD4;
+	color: #FFF;
+
+	&:hover {
```
should have `&:focus` here as well. however, i have a PR currently open that addresses us styling this button over and over again:

https://github.com/moderntribe/events-gutenberg/pull/320/files#diff-45709bc27430fe9a9fb02d309e4e1264R18

we should be using the class `tribe-editor__button--sm` for everything where we have the small blue button from the style guide.

mitogh replied:
Great adding a TODO here instead to remove this and update for a global style instead.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/footer/style.pcss:
```diff
@@ -0,0 +1,28 @@
+		color: #FFF;
+	}
+
+	&:disabled {
```
you can simplify this by doing:

```
&:disabled {
  &,
  &:hover,
  &:focus {
    /* your styles go here */
  }
}
```

mitogh replied:
Curios how we set the `disabled` and `focus` style here if both are different.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/header/template.js:
```diff
@@ -0,0 +1,115 @@
+			</div>
+			<div className="tribe-editor__new-ticket__content">
+				<AutosizeInput
+					id="ticket-creation-title"
```
this should be using a unique id. it's likely there won't be 2 ticket creations happening at the same time, but we cannot have 2 of the same id on the page.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/header/template.js:
```diff
@@ -0,0 +1,115 @@
+					onChange={ sendValue( setTitle ) }
+				/>
+				<AutosizeInput
+					id="ticket-creation-description"
```
same comment as above on the id

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/header/style.pcss:
```diff
@@ -0,0 +1,99 @@
+	border: 1px solid #AEB4BB;
+	border-bottom: none;
+
+	&:after {
```
we should not be using spacer elements, even if they are pseudo elements, instead, use a margin to the right of `.tribe-editor__new-ticket__price`.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/header/style.pcss:
```diff
@@ -0,0 +1,99 @@
+	display: flex;
+	flex-wrap: wrap;
+
+	.tribe-editor__new-ticket__title,
```
these shouldn't need to be under `.tribe-editor__new-ticket__content` as we are unnecessarily increasing the specificity of these elements, the BEM should put it in its own 'namespace', so to speak, and should be enough.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/sku/template.js:
```diff
@@ -0,0 +1,53 @@
+
+const SKU = ( { onChange, value, label, tooltip } ) => (
+	<div className="tribe-editor__tickets-form__row">
+		<div className="tribe-editor__tickets-form__labels">
```
not wrong, but these shouldn't need these wrapper divs.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/sku/template.js:
```diff
@@ -0,0 +1,53 @@
+	<div className="tribe-editor__tickets-form__row">
+		<div className="tribe-editor__tickets-form__labels">
+			<LabelWithTooltip
+				id="ticket-sku"
```
these ids should be unique, as mentioned in previous comments.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/sku/template.js:
```diff
@@ -0,0 +1,53 @@
+			/>
+		</div>
+		<div className="tribe-editor__tickets-form__input-group">
+			<input id="ticket-sku" type="text" value={ value } onChange={ sendValue( onChange ) } />
```
unique id

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/capacity/template.js:
```diff
@@ -0,0 +1,153 @@
+					selected={ type }
+					options={ capacityOptions }
+					onSelect={ onSelectType }
+					id="tickets-capacity-type"
```
same thing with unique id

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/capacity/template.js:
```diff
@@ -0,0 +1,153 @@
+			<label htmlFor={ `ticket-input-${ type }` }>{ input.label }</label>
+			<input
+				type="number"
+				id={ `ticket-input-${ type }` }
```
This id needs to be unique. In the `Input` elements below, they are just passing in a `type`, which creates 2 inputs with the same id. This is also the case with multiple ticket blocks, and you have multiple elements with the same id.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/header/template.js:
```diff
@@ -0,0 +1,115 @@
+	const priceInput = (
+		<AutosizeInput
+			key="price-input"
+			id="ticket-creation-description"
```
unique id

mitogh commented on plugins/tickets/src/modules/blocks/tickets/settings/template.js:
```diff
@@ -0,0 +1,21 @@
+const Settings = () => (
+	<Fragment>
+		<CapacityTable />
+		<HeaderImage />
```
For now is just a placeholder with no toggle capability, mainly to group all this elements once that one is available.

mitogh commented on plugins/tickets/src/modules/blocks/tickets/style.pcss:
```diff
@@ -2,3 +2,34 @@
 	color: #545D66;
 	font-family: Helvetica, sans-serif, arial;
 }
+
```
Seems like was added after the merge conflicts will double check.

mitogh commented on plugins/tickets/src/modules/elements/accordion-header/element.js:
```diff
@@ -0,0 +1,36 @@
+ */
+import { Dashicon } from '@wordpress/components';
+
+const AccordionHeader = ( { active, title, iconClass, titleClass } ) => (
```
Right however the idea is to remove the one from RSVP instead and use this one instead to decrease the amount of components with the same behavior across multiple elements.

paulmskim replied:
As we build elements on top of each other to create components and blocks, things become more and more specific. At the block level is where specificity is supposed to be, not at the element level. The elements are supposed to provide a generic template and interface for which to build blocks with, which provide their own specific requirements.

Like I said, technically speaking this isn't wrong, but semantically speaking, this is way too specific for an element.

mitogh replied:
In this particular case I think it does makes sense as you are creating the same `header` for both different containers, unless we have a single container for `accordions` as both `accordions` share the same structure in design and markup. 

I think having a single one will help to have similar patterns from design group as components just like the `small btn` as is the same one used in multiple places, the same style and structure for the header is used in this two accordions.

mitogh replied:
Looking into more detail I think you're right, removing this one 👍

mitogh commented on plugins/tickets/src/modules/elements/label-with-tooltip/element.js:
```diff
@@ -11,20 +11,38 @@ import classNames from 'classnames';
 import { Tooltip } from '@moderntribe/common/elements';
 import './style.pcss';
 
+const Label = ( { id, label } ) => {
```
I think this one of the things we discuss on the call yesterday to split this one in order to render a `<label>` instead.

paulmskim replied:
So I don't think I understood you when you said you were going to use a label. I'm pretty sure when we discussed this on the call, I was of the mind that using a label within this component is not the best idea.

As mentioned above, I don't think using a label component is best here. I'd love to discuss if you think otherwise.

mitogh replied:
I think this will prevent us to create another component just to use a `<label>` tag instead, there might be cases when just a `<span>` is good but as long as you pass and ID to associated with the `<label>` the component will use a `<label>`. 

Probably create another component for this seems to more complex from my point of view as both are going to share same structure only difference is one is using `<span>` and other `<label>`

mitogh commented on plugins/tickets/src/modules/elements/ticket-form/advanced-options/style.pcss:
```diff
@@ -0,0 +1,44 @@
+.tribe-editor__tickets__advanced-options {
+	.tribe-editor__accordion__row-header {
+		background-color: transparent;
```
I think this one relates with the one above when we have the styles in multiple places for the same style of component so I think having a single style for the component would be ideal.

Will update the styles of this one in the location of the component, instead.

mitogh commented:
I think I've addressed all the comments and answer the rest, thanks for such a great review.

Let me know if you have more concerns / questions to be addressed before moving this one into the `feature/` branch.

Thanks @paulmskim

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/header/template.js:
```diff
@@ -42,7 +43,7 @@ const Header = ( props ) => {
 		<AutosizeInput
 			key="price-input"
-			id="ticket-creation-description"
+			id={ uniqid( 'ticket-creation-price-' ) }
```
this component needs to be a stateful component that extends `PureComponent`, much like you did here:

https://github.com/moderntribe/events-gutenberg/pull/319/commits/ea10a086c07e9d1f1a0a33ad6e7ff5c7624e49fd

If any of these props change, because this component is stateless, the entire component will rerender, which in turn will call `uniqid()` again for every instance, recreating a new unique id.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/header/template.js:
```diff
@@ -59,15 +60,15 @@ const Header = ( props ) => {
 			<div className="tribe-editor__new-ticket__content">
 				<AutosizeInput
-					id="ticket-creation-title"
+					id={ uniqid( 'ticket-creation-title-' ) }
```
same on the id comment above

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/header/template.js:
```diff
@@ -59,15 +60,15 @@ const Header = ( props ) => {
 				/>
 				<AutosizeInput
-					id="ticket-creation-description"
+					id={ uniqid( 'ticket-creation-description-' ) }
```
same on the id comment above

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/capacity/template.js:
```diff
@@ -42,19 +43,22 @@ Select.propTypes = {
-	)
-);
+const Input = ( { type, input, selected } ) => {
+	const id = uniqid( `ticket-input-${ type }-` );
```
we should be passing in the id as a prop instead of defining it within the component. also, if any props change, this entire component will rerender, creating a new unique id.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/capacity/template.js:
```diff
@@ -77,11 +81,13 @@ const Capacity = ( props ) => {
 		shared
 	} = props;
+
+	const selectID =  uniqid( 'capacity-type-' );
```
this `Capacity` component should be a stateful component extending `PureComponent`, much like you did here: https://github.com/moderntribe/events-gutenberg/pull/319/commits/ea10a086c07e9d1f1a0a33ad6e7ff5c7624e49fd

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/sku/template.js:
```diff
@@ -10,28 +10,32 @@ import noop from 'lodash/noop';
-				tooltipLabel={ <Dashicon icon="info-outline" /> }
-			/>
+const SKU = ( { onChange, value, label, tooltip } ) => {
+	const id = uniqid( 'ticket-sku' );
```
same note on the unique id as above. this needs to extend `PureComponent`.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/header/template.js:
```diff
@@ -0,0 +1,116 @@
+	const priceInput = (
+		<AutosizeInput
+			key="price-input"
+			id={ uniqid( 'ticket-creation-price-' ) }
```
the `Header` component needs to be a stateful component that extends `PureComponent`, much like you did here: https://github.com/moderntribe/events-gutenberg/pull/319/commits/ea10a086c07e9d1f1a0a33ad6e7ff5c7624e49fd

If any of these props change, because this component is stateless, the entire component will rerender, which in turn will call `uniqid()` again for every instance, recreating a new unique id.

mitogh replied:
Right, however in this case the new ID does not make any difference as is just an associated with the label, so having a different ID will not have any difference here as having the same as before.

mitogh replied:
Would you mind take a look at this @paulmskim ?

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/header/template.js:
```diff
@@ -0,0 +1,116 @@
+			</div>
+			<div className="tribe-editor__new-ticket__content">
+				<AutosizeInput
+					id={ uniqid( 'ticket-creation-title-' ) }
```
same comment as above on `uniqid` and stateful component.

mitogh replied:
Same applied here.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/header/template.js:
```diff
@@ -0,0 +1,116 @@
+					onChange={ sendValue( setTitle ) }
+				/>
+				<AutosizeInput
+					id={ uniqid( 'ticket-creation-description-' ) }
```
same as above on `uniqid` and stateful component.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/capacity/template.js:
```diff
@@ -0,0 +1,158 @@
+
+// Custom input for this type of form
+const Input = ( { type, input, selected } ) => {
+	const id = uniqid( `ticket-input-${ type }-` );
```
we should be passing in the id as a prop instead of defining it within the component. also, if any props change, this entire component will rerender, creating a new unique id.

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/capacity/template.js:
```diff
@@ -0,0 +1,158 @@
+		shared
+	} = props;
+
+	const selectID =  uniqid( 'capacity-type-' );
```
this `Capacity` component should be a stateful component extending `PureComponent`, much like you did here: https://github.com/moderntribe/events-gutenberg/pull/319/commits/ea10a086c07e9d1f1a0a33ad6e7ff5c7624e49fd

paulmskim commented on plugins/tickets/src/modules/elements/ticket-form/sku/template.js:
```diff
@@ -0,0 +1,56 @@
+import { LabelWithTooltip } from '@moderntribe/tickets/elements';
+
+const SKU = ( { onChange, value, label, tooltip } ) => {
+	const id = uniqid( 'ticket-sku' );
```
same note on the unique id as above. this needs to extend `PureComponent`.

paulmskim commented:
@mitogh still have a couple here: https://github.com/moderntribe/events-gutenberg/pull/319#discussion_r218719427 and https://github.com/moderntribe/events-gutenberg/pull/319#discussion_r218719552

also left you a note on https://github.com/moderntribe/events-gutenberg/pull/319#discussion_r218932676

mitogh commented:
I think all items has been addressed.

Thanks and let me know if you have more questions.
